### ISSUE TITLE
Simple MLP breaks

### ISSUE THREAD

**Issue by botev** [clab/dynet#345]
Sorry, as this might be easy question, but I'm quite new to DyNet. I tried the following:
```
    w = model.parameters_from_numpy(xavier_normal_init(shape))
    b = model.parameters_from_numpy(constant_init((shape[1], )))
    w = dy.parameter(w)
    b = dy.parameter(b)
    f = nonlinearity_from_string(nonlinearity)
    return f(x_in * w + b)
```
and it gives me:
```
terminate called after throwing an instance of 'std::invalid_argument'
  what():  Mismatched input dimensions in Sum: [{64,64} {64}]
```
This seem like a basic thing so can someone point me to what I'm doing wrong. The x_in is a matrix however.

**Comment by botev** [comment]
Ok so from what I've learn DyNet has its own mechanism for batching, thus my input should be a vector. However, now I'm having problem finding how to provide minibatches at runtime - e.g. I want to run it as a static graph similar to Theano and tensorflow. I'm having difficulties in 1. defining an integer input, also finding how to take a mean over the minibatch (not sum).
E.g. at the end after a linear layer I have:
```
y = dy.scalarInput(0)
cost = dy.sum_batches(dy.pickneglogsoftmax(h, y))
```
Currently, DyNet complains that y is not an integer, but I don't know how to make it an integer, as well as how to take the mean of the cost not the sum.

**Comment by pmichel31415** [comment]
Not sure I understood everything, here's how it works : 

First bug : 

From what I understand you want `x_in` to be batched right? You can read more about minibatching in the docs but basically what you want is : 

    x_vals=np.zeros((64,64))
    x_in = dy.inputTensor(x_vals,batched = True) # The last dimension of x_vals 
                                                 # will be used as a batch dimension 
    w = model.parameters_from_numpy(xavier_normal_init(shape))
    b = model.parameters_from_numpy(constant_init((shape[1], )))
    w = dy.parameter(w)
    b = dy.parameter(b)
    f = nonlinearity_from_string(nonlinearity)
    return f(x_in * w + b)

Second bug : 

The second argument of `dy.pickneglogsoftmax` should just be a (python) integer so 

    y = 0
    cost = dy.sum_batches(dy.pickneglogsoftmax(h, y))

should work.

There is no specific operation to average over a batch but you can just sum and then divide by the batch size.

---

I recommend you take a look at the python tutorial [here](http://dynet.readthedocs.io/en/latest/tutorial.html#python-tutorial). I will be working on documenting the python bindings of dynet shortly so hopefully by next week you'll be able to open a python console and type `help(...)` to get more info about any function.

**Comment by botev** [comment]
Hmm, so what is the difference between `dy.inputTensor` with a provided numpy value and `dy.matInput` with only specified dimensions?

I'm trying to benchmark DyNet and it will be on a static function so I want to implement it such that it is as optimal as possible. E.g. at each iteration I will be passing new data. I was thinking this will happen similar to theano, but do I just need to reconstruct the whole expression every time with the new numpy inputs?

**Comment by pmichel31415** [comment]
So `dy.inputTensor` only takes a numpy ndarray (vector, matrix, tensor...) and infers the dimensions from the shape of the array (and with `batched=True` it uses the last dimension as a batch dimension) whereas matinput creates an empty matrix of the given dimensions.

Yes you need to recreate the graph each time before calling forward/backward. Basically you would have a function `getMLPLoss(x,y)` that builds the expression for your objective function, on which you'll then call `.scalar_value()` (forward) and `.backward()`.

There is more details in the paper https://arxiv.org/abs/1701.03980

**Comment by neubig** [comment]
Just an additional note: the C++ interface does support static declaration like theano, but I'm not sure if Python does. It might be one thing with considering.

**Comment by yoavg** [comment]
To add on @neubig 's comments:

1) Yes, you do need to create the graph each time, but this construction is very fast.
2) The Python bindings do support "static" graphs to some extent: you can construct a graph and then change just the input nodes. (I am not sure this holds for the inputTensor yet).
3) But you really shouldn't care much about doing (2), as creating the graph is really as fast once you start doing any "real" computation. Adapt your thinking to how DyNet works.
4) In particular, numpy ndarrays and DyNet internal matrices don't share the same memory, so there is a memory copy whenever you create a DyNet inputNode from a numpy array OR when you "set" the numpy array of an already existing node. And the memory copy is by far the dominating factor speed-wise in these operations.

**Comment by neubig** [comment]
Because the main question here has probably been resolved, I'll close this now in favor of #346

**Comment by botev** [comment]
just to clarify in that case @yoavg does DyNet has something similar to tensorflow queus or pytorch own arrays, which to allow more efficient moving of data?
